api/desecapi/utils.py

```diff
@@ -0,0 +1,14 @@
+from django.db.models import CharField, F, Value
+from django.db.models.functions import Concat, Length
+
+from desecapi.models import Domain
+
+
+def get_domain_for_qname(qname, **kwargs):
+    domain = Domain(name=qname)
+    domain.clean_fields(exclude=['owner'])  # syntactically validate domain name
```

Comment:
could break when we include more fields in `Domain` -- is there a way to define a positive-list instead of an exclude list?

Reply:
Unfortunately not.

Btw, why do this check at all, why worry about it here? If it's invalid, the database query will fail. -- No, because we're comparing via `_endswith`. It can happen that an invalid `qname` has a valid suffix, and we thus extract the wrong name. For example, if you own `example.com` and set `qname = 'admin@internal.example.com'`, it would match!

Reply:
There [is a way](https://github.com/desec-io/desec-stack/compare/20210127_id_manager#diff-53877dd549623d1da7ffa608733f27c53cb91a493ba8875226ddf8020cb45853R1108-R1112) to only use the domain name validation, not sure if that's more appropriate here.

Reply:
Also, technically, `admin@internal.example.com` is a valid qname, so I'm not sure how we should deal with this edge case

Reply:
Validation: fixed.

We allow logins with email addresses and token; allowing `@` in subdomains creates a uniqueness conflict here. Also, I'd say that dynDNS users won't usually have `@` in their dyn hostnames. No one has asked for this; I proposed to revisit once it becomes a problem.